platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java

```diff
@@ -302,6 +299,93 @@ private void writeMetadata(final Path dir) throws IOException {
+            final long bucketId = i;
+            final long bucketDataLocation = bucketIndexToBucketLocation.get(bucketId);
+            if (bucketDataLocation <= 0) {
+                continue;
+            }
+            final BufferedData bucketData =
+                    fileCollection.readDataItemUsingIndex(bucketIndexToBucketLocation, bucketId);
+            if (bucketData == null) {
+                throw new IOException("Cannot read bucket=" + bucketId + ", data location=" + bucketDataLocation);
+            }
+            try (final ParsedBucket bucket = new ParsedBucket()) {
+                bucket.readFrom(bucketData);
+                if (bucket.getBucketIndex() != bucketId) {
+                    logger.warn(MERKLE_DB.getMarker(), "Delete stale bucket: {}", bucketId);
+                    bucketIndexToBucketLocation.remove(bucketId);
+                    continue;
+                }
+                bucket.forEachEntry(entry -> {
+                    final Bytes keyBytes = entry.getKeyBytes();
+                    final long path = entry.getValue();
+                    try {
+                        boolean removeKey = true;
+                        if ((path < firstLeafPath) || (path > lastLeafPath)) {
+                            logger.warn(
+                                    MERKLE_DB.getMarker(), "Delete key (path range): key={}, path={}", keyBytes, path);
+                        } else {
+                            final BufferedData recordBytes = store.get(path);
+                            assert recordBytes != null;
+                            final VirtualLeafBytes record = VirtualLeafBytes.parseFrom(recordBytes);
+                            if (!record.keyBytes().equals(keyBytes)) {
+                                logger.warn(
+                                        MERKLE_DB.getMarker(),
+                                        "Delete key (stale): path={}, expected={}, actual={}",
+                                        path,
+                                        record.keyBytes(),
+                                        keyBytes);
+                            } else {
+                                assert record.path() == path;
+                                removeKey = false;
+                            }
+                        }
+                        if (removeKey) {
+                            delete(keyBytes, entry.getHashCode());
+                        }
+                    } catch (final Exception e) {
+                        logger.error(
+                                MERKLE_DB.getMarker(),
+                                "Exception while processing bucket entry, bucket={}, key={} path={}",
+                                bucketId,
+                                keyBytes,
+                                path,
+                                e);
+                    }
+                });
+            }
+        }
+        endWriting();
+        logger.info(
```

Comment:
We need to check if we have found all keys in the `[firstLeafPath..lastLeafPath]` range. If some keys are not found, we can't restore them (due to missing hashcodes) but we need to report that.

Reply:
This is responsibility of the state validation tool, I propose

Reply:
Done